### DISCUSSION TITLE
Exporting CFE API to library for Rust Bindgen

### DISCUSSION THREAD

**Discussion by tstern27** [nasa/cFS#504] [Q&A]
Hello, 

I am exploring attempting to integrate Rust into cFS through the use of Bindgen. I know this has been attempted a few times before me, but I have been running into issues with the integration, specifically finding the libraries containing the information I need. 

As a quick background Bindgen generates interface code for C->Rust, by supplying a list of headers and libraries containing the information desired. I have references the sample application to get my list of headers and now I have been trying to get the libraries pulled together. 

There are some static libs generated through cmake in cFE (libmsg.a, libevs.a, libconfig.a, etc.) that I am using however I still get errors of a missing symbol "GLOABL_CONFIGDATA" that is not found in any static library. It seems to be a part of some object files that generate the executable. 

Digging further into the cmake process I see that the cFS applications depend on an Interface library, core_api, created by cmake however this is not something that can be exported, best I can tell. 

My question is, is there a logical point to export the cFE core_api library so that it contains the interfaces needed for running a cFS application?

Any and all help is greatly appreciated. 
Thanks, 
Tim Stern

**Comment by skliper** [comment]
GLOBAL_CONFIGDATA comes from target_config.c: https://github.com/nasa/cFE/blob/10898c9b1292be87cd851fb58eeda253dfdc4358/cmake/target/src/target_config.c#L132

and it's getting built into the core executable here:
https://github.com/nasa/cFE/blob/10898c9b1292be87cd851fb58eeda253dfdc4358/cmake/target/CMakeLists.txt#L116-L125

The core_api interface library is really just leveraging CMake methods to add the necessary includes when building apps that depend on the APIs.  There's no object code associated with interface libraries.

**Comment by zec** [comment]
This isn't _quite_ what you were asking for, but:

I've recently released Rust bindings to (parts of) cFE and OSAL:

https://github.com/BlackCAT-CubeSat/n2o4

Even if you don't want to use the safe wrappers `n2o4` provides, check the `cfs-sys` and `etc` directories for how we use `bindgen` (with some CMake-ery and manual additions) to create the low-level, unsafe FFI binding.
